### ISSUE TITLE
Prediction Intervals

### ISSUE THREAD

**Issue by AhmedThahir** [jkitchin/pycse#34]
This is regarding

https://github.com/jkitchin/pycse/blob/329733b75fc7ba01215ff64a56cd7b2f3f0a9b32/pycse/PYCSE.py#L237

It looks like this is the formula for confidence interval for the mean response. The prediction intervals has an extra $+1$ term

https://github.com/jkitchin/pycse/blob/329733b75fc7ba01215ff64a56cd7b2f3f0a9b32/pycse/PYCSE.py#L274

Please correct me if I'm wrong:

![image](https://github.com/jkitchin/pycse/assets/94914628/6b9cc8a9-c42b-488a-a826-48d65823f47e)

**Comment by jkitchin** [comment]
Do you have a reference for this image? I can see what you are talking about. The push above gets it more correct, but still leaves out the CovMatrix term.

**Comment by AhmedThahir** [comment]
1. Shouldn't the (1+1/n) in the following line be np.sqrt(1+1/n)

https://github.com/jkitchin/pycse/blob/ac526683d1fb9d0ba115115251d7f8f9f030b9e6/pycse/PYCSE.py#L203

This is referring to the GitHub page: https://en.wikipedia.org/wiki/Prediction_interval#Unknown_mean,_unknown_variance
![Image](https://github.com/user-attachments/assets/0f3f407d-d37a-4db0-b222-3b410c60f22f)

If yes, I would like to make a PR for it

2. Regarding the references

- Univariate Linear Regression: https://www.youtube.com/watch?v=tJ8cv-fFPHs
- Multiple Linear Regression: https://www.youtube.com/watch?v=6TYyb3JrtWw
- Multivariate non-linear regression: your approach


<img alt="Image" src="https://github.com/user-attachments/assets/dc35a3e0-9660-43a8-9745-1260aaef42a6" />

**Comment by AhmedThahir** [comment]
1. Shouldn't the (1+1/n) in the following line be np.sqrt(1+1/n)

https://github.com/jkitchin/pycse/blob/ac526683d1fb9d0ba115115251d7f8f9f030b9e6/pycse/PYCSE.py#L203

This is referring to the GitHub page: https://en.wikipedia.org/wiki/Prediction_interval#Unknown_mean,_unknown_variance
![Image](https://github.com/user-attachments/assets/0f3f407d-d37a-4db0-b222-3b410c60f22f)

If yes, I would like to make a PR for it

2. Regarding the references

- Univariate Linear Regression: https://www.youtube.com/watch?v=tJ8cv-fFPHs
- Multiple Linear Regression: https://www.youtube.com/watch?v=6TYyb3JrtWw
- Multivariate non-linear regression: your approach

<img alt="Image" src="https://github.com/user-attachments/assets/f07d300e-0343-4c95-889e-b47b1637df4c" />

> Do you have a reference for this image? I can see what you are talking about. The push above gets it more correct, but still leaves out the CovMatrix term.

From my understanding, the CovMatrix is just $(X'X)^{-1}$

**Comment by AhmedThahir** [comment]
It's odd for me that the simple linear regression, multi-variate linear regression and non-linear regression are not similar in the part after the RMSE x

I would like your feedback on these references

**Comment by AhmedThahir** [comment]
This is what I derived from all the resources. Please verify:

<img width="1245" alt="Image" src="https://github.com/user-attachments/assets/59ba5a4c-fb67-4a4f-8f89-cafab5d57df0" />

**Comment by AhmedThahir** [comment]
This is what I derived from all the resources. Please verify:

<img width="1245" alt="Image" src="https://github.com/user-attachments/assets/284d58ea-d0ea-4869-830b-436861947754" />

**Comment by AhmedThahir** [comment]
This is what I derived from all the resources. Please verify:

<img width="1245" alt="Image" src="https://github.com/user-attachments/assets/4f135932-bc5c-4767-bf49-b31edb389430" />

**References**

- Multiple Linear Regression: https://www.youtube.com/watch?v=6TYyb3JrtWw
- Multivariate non-linear regression: Multiple Linear Regression modified using Delta Approach

---

> Do you have a reference for this image? I can see what you are talking about. The push above gets it more correct, but still leaves out the CovMatrix term.

From my understanding, the CovMatrix is just $(X'X)$

**Comment by AhmedThahir** [comment]
I would like to know why hat matrix calculation has a $2 \times$:

https://github.com/jkitchin/pycse/blob/ac526683d1fb9d0ba115115251d7f8f9f030b9e6/pycse/PYCSE.py#L179

**Comment by jkitchin** [comment]
The 2 comes from minimizing the summed squared error, and not 1/2 the summed squared error. That 2 is part of the derivative of the loss function. If you minimize 1/2 SSE, then the 1/2 cancels the 2.

**Comment by jkitchin** [comment]
> This is what I derived from all the resources. Please verify:
> 
> ![Image](https://github.com/user-attachments/assets/34c9ab3b-61db-4884-b5ef-152893daa127)
> 
> **References**
> 
> * Simple Linear Regression: https://www.youtube.com/watch?v=3CHDD_ezgn8
> * Simple Linear Regression: https://www.youtube.com/watch?v=tJ8cv-fFPHs
> * Multiple Linear Regression: https://www.youtube.com/watch?v=6TYyb3JrtWw
> * Multivariate non-linear regression: Multiple Linear Regression modified using Delta Approach
> 
> > Do you have a reference for this image? I can see what you are talking about. The push above gets it more correct, but still leaves out the CovMatrix term.
> 
> From my understanding, the CovMatrix is just ( X ′ X )

These look approximately right, but the use of H ~ J.T@J is only right for near linear systems, and the use of a full Hessian is better imo, since it is easy to get. When you use the full hessian, it appears you don't need the extra 1/n terms, and maybe not the 1 term. either as those seem to be included in the linear algebra.

In general it is my conclusion from looking at lots of sources that all of these are approximately right, and use different approximations for some convenience. That means they are also all approximately wrong too. There are a lot of details here and there that determine if you need a factor of 2 or not, and when you can approximate a hessian by a dot-product or not.

**Comment by AhmedThahir** [comment]
> These look approximately right, but the use of H ~ J.T@J is only right for near linear systems, and the use of a full Hessian is better imo, since it is easy to get. When you use the full hessian, it appears you don't need the extra 1/n terms, and maybe not the 1 term. either as those seem to be included in the linear algebra.

Understood.

> In general it is my conclusion from looking at lots of sources that all of these are approximately right, and use different approximations for some convenience. That means they are also all approximately wrong too. There are a lot of details here and there that determine if you need a factor of 2 or not, and when you can approximate a hessian by a dot-product or not.

So, wouldn't it be better to take the Hessian and Jacobian using autograd (like you showed in one of your research videos) instead of using XX and $(X'X)^{-1}$?

**Comment by AhmedThahir** [comment]
> The 2 comes from minimizing the summed squared error, and not 1/2 the summed squared error. That 2 is part of the derivative of the loss function. If you minimize 1/2 SSE, then the 1/2 cancels the 2.

Got it.
Doesn't minimizing SSE and 1/2 SSE return the same optimization? As multiplying by 2 is a constant linear monotonic transformation.

**Comment by AhmedThahir** [comment]
On a sidenote: it's great of you for taking your time to have this discussion with me. Really appreciate it!

**Comment by jkitchin** [comment]
> > These look approximately right, but the use of H ~ J.T@J is only right for near linear systems, and the use of a full Hessian is better imo, since it is easy to get. When you use the full hessian, it appears you don't need the extra 1/n terms, and maybe not the 1 term. either as those seem to be included in the linear algebra.
> 
> Understood.
> 
> > In general it is my conclusion from looking at lots of sources that all of these are approximately right, and use different approximations for some convenience. That means they are also all approximately wrong too. There are a lot of details here and there that determine if you need a factor of 2 or not, and when you can approximate a hessian by a dot-product or not.
> 
> So, wouldn't it be better to take the Hessian and Jacobian using autograd (like you showed in one of your research videos) instead of using XX and ( X ′ X ) − 1 ?

in general yes, and in the nlpredict we use numdifftools for that. autograd itself is outdated, and I thought jax was too heavy to make a requirement, and it won't even always work if you don't know what you are doing. 

for linear regression you either make the assumption of the loss or also use something like numdifftools (which requires the loss function). There isn't a perfect answer imo. At least in linear regression it is only off by a factor of 2 at worst I think.

**Comment by jkitchin** [comment]
> > The 2 comes from minimizing the summed squared error, and not 1/2 the summed squared error. That 2 is part of the derivative of the loss function. If you minimize 1/2 SSE, then the 1/2 cancels the 2.
> 
> Got it. Doesn't minimizing SSE and 1/2 SSE return the same optimization? As multiplying by 2 is a constant linear monotonic transformation.

it does return the same optimization parameters, but it changes the sensitivity of the loss. It is only linear, but in the delta method, that sensitivity is related to the uncertainty.

**Comment by jkitchin** [comment]
> On a sidenote: it's great of you for taking your time to have this discussion with me. Really appreciate it!

no problem. This is the stickiest and hardest thing in pycse, and it has been hard finding definitive references imo that are easy to understand and implement. I appreciate your interest.

**Comment by AhmedThahir** [comment]
> in general yes, and in the nlpredict we use numdifftools for that. autograd itself is outdated, and I thought jax was too heavy to make a requirement, and it won't even always work if you don't know what you are doing.

1. Why is autograd outdated? It had a recent release a few months ago.
2. Jax supports Jax.numpy which is a drop-in replacement, with all the perks of jax auto differentiation

**Comment by jkitchin** [comment]
both of them are limited in what they can differentiate. If you put many scipy functions in your loss function (e.g. from your model), for example, it does not know how to differentiate through many of them. numdifftools on the other hand is basically sophisticated finite differences that might be slower, but more reliably works.

jax is the replacement for autograd.

**Comment by AhmedThahir** [comment]
> If you put many scipy functions in your loss function (e.g. from your model), for example, it does not know how to differentiate through many of them.

Though limited, Jax supports scipy.jax. wouldn't that cover most of the cases?

> numdifftools on the other hand is basically sophisticated finite differences that might be slower, but more reliably works.

Oh, that's something new for me! I always thought that analytical derivatives are better.

**Comment by AhmedThahir** [comment]
> no problem. This is the stickiest and hardest thing in pycse, and it has been hard finding definitive references imo that are easy to understand and implement. I appreciate your interest.

Glad to do so 😄

When I was working on non-linear optimization for my R&D internship at Reckitt-Benckiser, this is what I was stuck on as well. I made a great tool for modeling chemical kinetics of products, but I was unable to quantify the uncertainty. Till this day, I keep in touch with my manager hoping to get that sorted for them 😄

**Comment by jkitchin** [comment]
> > If you put many scipy functions in your loss function (e.g. from your model), for example, it does not know how to differentiate through many of them.
> 
> Though limited, Jax supports scipy.jax. wouldn't that cover most of the cases?
> 
> > numdifftools on the other hand is basically sophisticated finite differences that might be slower, but more reliably works.
> 
> Oh, that's something new for me! I always thought that analytical derivatives are better.

the point is i don't want to cover most of the cases, i want to cover all the cases. jax.scipy does not cover all of scipy, and you cannot rely on users to define everything with jax.numpy. You can't mix these either.

**Comment by AhmedThahir** [comment]
> numdifftools on the other hand is basically sophisticated finite differences that might be slower, but more reliably works.

Could you share any courses/references to finite differences. I'd like to learn more about it.

> the point is i don't want to cover most of the cases, i want to cover all the cases. jax.scipy does not cover all of scipy, and you cannot rely on users to define everything with jax.numpy. You can't mix these either.

Right! Makes sense. Thanks

**Comment by jkitchin** [comment]
> > numdifftools on the other hand is basically sophisticated finite differences that might be slower, but more reliably works.
> 
> Could you share any courses/references to finite differences. I'd like to learn more about it.

https://en.wikipedia.org/wiki/Numerical_differentiation

> 
> > the point is i don't want to cover most of the cases, i want to cover all the cases. jax.scipy does not cover all of scipy, and you cannot rely on users to define everything with jax.numpy. You can't mix these either.
> 
> Right! Makes sense. Thanks
